zbus/src/connection/mod.rs

```diff
@@ -1205,7 +1205,7 @@ impl Connection {
     /// Returns a listener, notified on various connection activity.
     ///
     /// This function is meant for the caller to implement idle or timeout on inactivity.
-    pub fn monitor_activity(&self) -> EventListener {
+    pub fn monitor_activity(&self) -> Pin<Box<EventListener>> {
```

Comment:
Yeah, this is public-only API. I tried this but now a test timeouts

```
---- blocking::connection::tests::unix_p2p stdout ----
thread 'blocking::connection::tests::unix_p2p' panicked at zbus/src/blocking/connection/mod.rs:292:5:
timeout: the function call took 15000 ms. Max time 15000 ms
```

```diff
diff --git a/zbus/src/blocking/connection/mod.rs b/zbus/src/blocking/connection/mod.rs
index ff4209d5..859d45f3 100644
--- a/zbus/src/blocking/connection/mod.rs
+++ b/zbus/src/blocking/connection/mod.rs
@@ -3,7 +3,7 @@
 use enumflags2::BitFlags;
 use event_listener::EventListener;
 use static_assertions::assert_impl_all;
-use std::{io, ops::Deref, pin::Pin};
+use std::{io, ops::Deref};
 use zbus_names::{BusName, ErrorName, InterfaceName, MemberName, OwnedUniqueName, WellKnownName};
 use zvariant::ObjectPath;
 
@@ -241,7 +241,7 @@ impl Connection {
     /// Returns a listener, notified on various connection activity.
     ///
     /// This function is meant for the caller to implement idle or timeout on inactivity.
-    pub fn monitor_activity(&self) -> Pin<Box<EventListener>> {
+    pub fn monitor_activity(&self) -> EventListener {
         self.inner.monitor_activity()
     }
 
@@ -313,7 +313,7 @@ mod tests {
         });
 
         let c = Builder::unix_stream(p1).p2p().build().unwrap();
-        let mut listener = c.monitor_activity();
+        let mut listener = Box::pin(c.monitor_activity());
         let mut s = MessageIterator::from(&c);
         tx.send(()).unwrap();
         let m = s.next().unwrap().unwrap();
@@ -326,10 +326,12 @@ mod tests {
         assert_eq!(val, "yay");
 
         // there was some activity
+        listener.as_mut().listen();
         listener.as_mut().wait();
         // eventually, nothing happens and it will timeout
         loop {
-            let mut listener = c.monitor_activity();
+            let mut listener = Box::pin(c.monitor_activity());
+            listener.as_mut().listen();
             if listener
                 .as_mut()
                 .wait_timeout(std::time::Duration::from_millis(10))
diff --git a/zbus/src/connection/mod.rs b/zbus/src/connection/mod.rs
index ac5a6352..55da7181 100644
--- a/zbus/src/connection/mod.rs
+++ b/zbus/src/connection/mod.rs
@@ -1205,8 +1205,8 @@ impl Connection {
     /// Returns a listener, notified on various connection activity.
     ///
     /// This function is meant for the caller to implement idle or timeout on inactivity.
-    pub fn monitor_activity(&self) -> Pin<Box<EventListener>> {
-        self.inner.activity_event.listen()
+    pub fn monitor_activity(&self) -> EventListener {
+        EventListener::new(&self.inner.activity_event)
     }
 
     /// Returns the peer credentials.
```

Reply:
> Yeah, this is public-only API. I tried this but now a test timeouts

That's very strange. I don't see how they can be connected? Perhaps it's just some timing issue?

Reply:
> Perhaps it's just some timing issue?

Thanks for the idea. Yes! `listener.as_mut().listen();` should be called earlier.

Reply:
Now different tests timeout...

```
---- iface_and_proxy_unix_p2p stdout ----
thread 'zbus::Connection executor' panicked at /home/maxv/.cargo/registry/src/index.crates.io-6f17d22bba15001f/event-listener-3.0.0/src/lib.rs:1069:36:
listener was never inserted into the list
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'iface_and_proxy_unix_p2p' panicked at zbus/tests/e2e.rs:688:1:
timeout: the function call took 15000 ms. Max time 15000 ms

---- iface_and_proxy stdout ----
thread 'zbus::Connection executor' panicked at /home/maxv/.cargo/registry/src/index.crates.io-6f17d22bba15001f/event-listener-3.0.0/src/lib.rs:1069:36:
listener was never inserted into the list
thread 'iface_and_proxy' panicked at zbus/tests/e2e.rs:681:1:
timeout: the function call took 15000 ms. Max time 15000 ms
```

Reply:
> Now different tests timeout...

:(

I'll look later this week if you don't manage to figure it out by then. BTW, I'd suggest first running tests locally to get a quicker feedback.

Reply:
I do run them locally. I will push current (broken) code though so you can reproduce the timeout.